Better engine, Cleaning and first implementations ...

Cranc opened the pull request:
- Compare main functions for SqlTypes
- Basic Compare functions
- Value Comparison for Int (i32)
- Value Comparison for String
- test function in table_test
- Documentation
- lookup
- fullscan
- insert_row
- more

Commit:
Add some more unittests

Added some further testing regarding select statements including where clauses.
Added tests for errors regarding create, alter, describe and use statements

Commit:
Add functionality to DataSrc changes to Rows usage of Primary key and more

Add get_row_with_primary_key
Add check_for_prmary_key
to flatfile engine
Add is_true for DataSrc
Add static to_bool for converting u8 to bool
Add some FromSql with NoImplementaion
Add get_value_by_name to Row with NoImplementation

Commit:
Fix string literal \n

mflueggen committed:
Implement FromSql Trait for String, bool, i32 and u16 and table.get_description()

mflueggen committed:
Fix style check

Commit:
Merge branch dev of mflueggen full implementation of primary key search

Workin structure
Missing way to transfer data into copy of rows

SvantjeJung committed:
Add more convienent and readable output of Rows on console

SvantjeJung committed:
Merge branch 'dev' of https://github.com/OsnaCS/uosql-server into dev

Conflicts:
	server/src/conn.rs

Commit:
solve some logic errors

Commit:
cleaning rows

Commit:
Merge branch 'better_engine' of github.com:rieping/uosql-server into better_engine

SvantjeJung committed:
Add display meta-data in client

SvantjeJung committed:
Tiny fix

Commit:
cleanup

SvantjeJung committed:
Merge pull request #74 from raya1234/dev

Add some more unittests

SvantjeJung committed:
Merge pull request #81 from SvantjeJung/dev

Display output of Rows on console

Commit:
Add cmp fn to SqlType and implemented compare_byte_for_equal

Commit:
Add compare_byte_greater_than and lesser_than

finished comp match tree
added basic compare methods with simple byte compare

Commit:
Add compare_as_bool fn for distinct bool equals compare

Commit:
Add Documentation

Commit:
Add specific attributes and subsequent functions for few methods.

Expanded column info by adding more attributes as options including default values.
Added functionality regarding parsing SQL statements that include the keyword "view".
Subsequent changes to unittest.

Now able to parse "Order By" attribute.

Commit:
old style-check error

Commit:
Add value comparison for Int

Commit:
Create template for Rows struct

Commit:
Thinking about next_row()

might implement BufReader

Commit:
Will probably use BufStream to read and write from buffer

Commit:
Add checks of function to test and implements more compare functions for string

Commit:
style check

Commit:
Fix branch issues and some snake case camel case errors.

Commit:
Changes function used in cmp

Commit:
Merge branch 'dev' of github.com:OsnaCS/uosql-server into dev

Commit:
Fix lines that are too long

Commit:
Remove warnings in Sqltype

Commit:
compilable version of data.rs

Commit:
Fix compile issues

Commit:
Merge branch 'better_engine' of https://github.com/rieping/uosql-server into better_engine

LukasKalbertodt commented on server/src/parse/ast.rs:
```diff
@@ -206,5 +206,41 @@ pub enum DataSrc  {
+                true
+            },
+        }
+    }
```
This method can be written much shorter:

```
pub fn is_true(&self) -> bool {
    match self {
        &DataSrc::Int(x) => x == 0,
        &DataSrc::String(ref x) => !x.is_empty(),
        &DataSrc::Bool(x) => x != 0,
    }
}
```

It should be clear, why this is possible...

LukasKalbertodt commented on server/src/parse/ast.rs:
```diff
@@ -206,5 +206,41 @@ pub enum DataSrc  {
+            return false
+        }
+        true
+    }
```
This method can be shorter in the same way as above

LukasKalbertodt commented on server/src/query.rs:
```diff
@@ -86,8 +86,9 @@ impl<'a> Executor<'a> {
-        try!(engine.insert_row(&n));
-        Ok(generate_rows_dummy())
+        // try!(engine.insert_row(&n));
+        return Err(ExecutionError::DebugError("engine.insert_row() not implemented ".into()));
```
If this is the last expression, the `result` may be omitted.

LukasKalbertodt commented on server/src/query.rs:
```diff
@@ -103,7 +104,8 @@ impl<'a> Executor<'a> {
         let engine = table.create_engine();
-        Ok(try!(engine.full_scan()))
+        // Ok(try!(engine.full_scan()))
+        return Err(ExecutionError::DebugError("engine.full_scan() not implemented ".into()));
```
No return again

LukasKalbertodt commented on server/src/storage/types.rs:
```diff
@@ -150,7 +132,279 @@ impl SqlType {
+                    },
+                    CompType::SEThan => {
+                        self.compare_byte_greater_than(val, val2).map(|x| !x)
+                    },
```
`> >= < <=` comparisons for bool are pretty useless... We would just disallow those queries.

LukasKalbertodt commented on server/src/storage/types.rs:
```diff
@@ -150,7 +132,279 @@ impl SqlType {
+            }
+        }
+        Ok(true)
+    }
```
Slices already implement the [`PartialEq`](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html) trait. So you can just compare `val` and `val2` with `==`. It will return `false` if the length does not match.

LukasKalbertodt commented on server/src/storage/types.rs:
```diff
@@ -150,7 +132,279 @@ impl SqlType {
+            }
+        }
+        Ok(false)
+    }
```
Same thing with [`PartialOrd`](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html). You can use `<`, `<=`, `>=` and `>` I think...

LukasKalbertodt commented on server/src/storage/types.rs:
```diff
@@ -150,7 +132,279 @@ impl SqlType {
+        }
+
+        Ok(false)
+    }
```
Just return an error if one of those slices is longer than one byte. Booleans are always one byte wide, so check length and compare with `==`

LukasKalbertodt commented on server/src/storage/types.rs:
```diff
@@ -150,7 +132,279 @@ impl SqlType {
+        if int1 != int2 {
+            return Ok(false)
+        }
+        Ok(true)
```
Just write:

``` rust
Ok(int1 == int2)
```

You have code like that in multiple functions.

LukasKalbertodt commented on server/src/storage/types.rs:
```diff
@@ -218,3 +476,30 @@ impl FromSql for u16 {
+            return Ok(false)
+        }
+        Ok(true)
+    }
```
Again, the function body is only:

``` rust
Ok(try!(data.read_u8()) != 0)
```

Commit:
Solves issues in pull request

Commit:
Implement basic seek functionality + next_row

Commit:
some code shorting

Commit:
Add function too set and is deleted in RowHeader

Commit:
style check

Commit:
solves error in query_example #1

Commit:
Merge branch 'better_engine' of https://github.com/Cranc/uosql-server into better_engine

Commit:
solves query example #2

Commit:
merge with cranc

Commit:
Merge branch 'better_engine' of https://github.com/rieping/uosql-server into better_engine

Commit:
add test

Commit:
add easter eggs

Commit:
minor bug fixes

Commit:
Fix minor bug

Commit:
implement lookup and full_scan into flatfile

Commit:
add delete_row

Commit:
Merge branch 'better_engine' of https://github.com/rieping/uosql-server into better_engine

Commit:
Add and implement Error::EndOfFile

Commit:
Merge branch 'better_engine' of https://github.com/rieping/uosql-server into better_engine

Commit:
Begin adding load function

Commit:
test for lookup final touches and more

Commit:
change get_value

SvantjeJung committed:
Add unique error messages for errors to display

Every error now has a short message to signal what went wrong. This can
be accessed via Error.description()

Commit:
Add lookup, fullscan and insert for row

fn is_empty for rows

SvantjeJung committed:
Merge pull request #84 from SvantjeJung/dev

Add unique error messages for errors to display

Commit:
Enable support for executing sql script

Commit:
merge with cranc and refactor lookup

Commit:
Enables full sql script functionality including commenting single lines

Commit:
Merge branch 'dev' of github.com:OsnaCS/uosql-server into dev

Commit:
Fix for error handling

SvantjeJung committed:
Merge pull request #85 from MaxDoll/dev

Enables loading sql script from file

SvantjeJung committed:
Merge pull request #83 from raya1234/dev

Add specific attributes and subsequent functions for few methods.

Commit:
add delete, add documentation, update table_test

Commit:
add documentation and last touches

Commit:
Add modify

Commit:
Merge branch 'dev' of https://github.com/OsnaCS/uosql-server into better_engine

some compiling errors

Commit:
Merge branch 'better_engine' of https://github.com/rieping/uosql-server into better_engine
